flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManager.java

```diff
@@ -0,0 +1,609 @@
+
+		// short-cut, pretty much only applicable to fine-grained resource management
+		if (pendingSlotProfiles.contains(resourceProfile)) {
+			pendingSlots.decrementCount(resourceProfile, 1);
+			return true;
+		}
+
+		for (ResourceProfile pendingSlotProfile : pendingSlotProfiles) {
+			if (pendingSlotProfile.isMatching(resourceProfile)) {
+				pendingSlots.decrementCount(pendingSlotProfile, 1);
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	// ---------------------------------------------------------------------------------------------
+	// Legacy APIs
+	// ---------------------------------------------------------------------------------------------
+
+	@Override
+	public int getNumberRegisteredSlots() {
+		return taskExecutorManager.getNumberRegisteredSlots();
+	}
+
+	@Override
+	public int getNumberRegisteredSlotsOf(InstanceID instanceId) {
+		return taskExecutorManager.getNumberRegisteredSlotsOf(instanceId);
+	}
+
+	@Override
+	public int getNumberFreeSlots() {
+		return taskExecutorManager.getNumberFreeSlots();
+	}
+
+	@Override
+	public int getNumberFreeSlotsOf(InstanceID instanceId) {
+		return taskExecutorManager.getNumberFreeSlotsOf(instanceId);
+	}
+
+	@Override
+	public Map<WorkerResourceSpec, Integer> getRequiredResources() {
+		return taskExecutorManager.getRequiredWorkers();
+	}
+
+	@Override
+	public ResourceProfile getRegisteredResource() {
+		return taskExecutorManager.getTotalRegisteredResources();
+	}
+
+	@Override
+	public ResourceProfile getRegisteredResourceOf(InstanceID instanceID) {
+		return taskExecutorManager.getTotalRegisteredResourcesOf(instanceID);
+	}
+
+	@Override
+	public ResourceProfile getFreeResource() {
+		return taskExecutorManager.getTotalFreeResources();
+	}
+
+	@Override
+	public ResourceProfile getFreeResourceOf(InstanceID instanceID) {
+		return taskExecutorManager.getTotalFreeResourcesOf(instanceID);
+	}
+
+	@Override
+	public void setFailUnfulfillableRequest(boolean failUnfulfillableRequest) {
+		// we always send notifications if we cannot fulfill requests, and it is the responsibility of the JobManager
+		// to handle it (e.g., by reducing requirements and failing outright)
+	}
+
+	@Override
+	public int getNumberPendingSlotRequests() {
+		// only exists for testing purposes
+		throw new UnsupportedOperationException();
+	}
+
+	// ---------------------------------------------------------------------------------------------
+	// Internal utility methods
+	// ---------------------------------------------------------------------------------------------
+
+	private void checkInit() {
+		Preconditions.checkState(started, "The slot manager has not been started.");
+	}
+
+	// ---------------------------------------------------------------------------------------------
+	// Testing methods
+	// ---------------------------------------------------------------------------------------------
+
+	@Override
+	@VisibleForTesting
+	public void unregisterTaskManagersAndReleaseResources() {
```

Comment:
Why is this method not implemented?

Reply:
This method only exists for a single test case in the `YarnResourceManagerTest`. I think it is quite questionable because it not only relies on RM-internal components (the SlotManager) to trigger the unregistration/release (instead of unregistering the TE at the RM), but also introduces a custom codepath into the SlotManager for releasing resources that in this form doesn't appear to exist in production.
The test asserts that the container is released upon calling this method, but it begs the question what we are even testing. If we are just interested in whether the Yarn RM resource actions can properly shut down the container, then we shouldn't need the slot manager for that.

Reply:
Can we refactor the test and get rid of this method then?